### PR TITLE
Implement Locate2Args & Locate3Args & LocateBinary2Args & LocateBinary3Args

### PR DESCRIPTION
**Pull request by gaodayue**
Signed-off-by: gaodayue <gaodayue@meituan.com>

## What have you changed? (mandatory)

implment builtin string function: Locate2Args & Locate3Args & LocateBinary2Args & LocateBinary3Args

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

**Comment by rleungx** [comment]
@gaodayue Thanks for your PR! 
The CI failed due to the code format. Could you format the code?

**Comment by gaodayue** [comment]
Thanks @rleungx , I have fixed the format issue. Could you take another look?

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 1]
```diff
@@ -60,6 +60,101 @@ impl ScalarFunc {
...
+        if substr.is_empty() {
+            return Ok(Some(pos + 1));
+        }
+
+        let slice = s
+            .char_indices()
+            .nth(pos as usize)
+            .map(|(offset, _)| &s[offset..]);
+
+        let result = match slice {
```
Comment (rleungx) [inline src/coprocessor/dag/expr/builtin_string.rs]: How about this?
```
        if let Some(slice) = slice {
            if let Some(idx) = slice.find(&substr) {
                return Ok(Some(1 + pos + slice[..idx].chars().count() as i64));
            }
        }
        Ok(Some(0))
```

Comment (gaodayue) [reply src/coprocessor/dag/expr/builtin_string.rs]: 👍

**Comment by gaodayue** [comment]
I think the CI failure is unrelated to this PR, right?

**Comment by siddontang** [comment]
PTAL @breeswish

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 2]
```diff
@@ -953,6 +1046,182 @@ mod tests {
...
+            let substr = Datum::Bytes(substr.as_bytes().to_vec());
+            let s = Datum::Bytes(s.as_bytes().to_vec());
+            let got = eval_func(ScalarFuncSig::Locate2Args, &[substr, s]).unwrap();
+            assert_eq!(got, Datum::I64(exp));
+        }
+
+        let null_cases = vec![
+            (Datum::Null, Datum::Bytes(b"foobar".to_vec()), Datum::Null),
+            (Datum::Bytes(b"bar".to_vec()), Datum::Null, Datum::Null),
+            (Datum::Null, Datum::Null, Datum::Null),
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_string.rs]: Could you add cases like `SELECT LOCATE('',NULL);SELECT LOCATE(NULL,'');`?

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 3]
```diff
@@ -953,6 +1046,182 @@ mod tests {
...
+            let got = eval_func(ScalarFuncSig::Locate2Args, &[substr, s]).unwrap();
+            assert_eq!(got, exp);
+        }
+    }
+
+    #[test]
+    fn test_locate_binary_2_args() {
+        let cases = vec![
+            ("", "foobArbar", 1),
+            ("", "", 1),
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_string.rs]: how about case `locate_binary("xxx","")`

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 2 above)
Comment (gaodayue) [reply src/coprocessor/dag/expr/builtin_string.rs]: sure

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 3 above)
Comment (gaodayue) [reply src/coprocessor/dag/expr/builtin_string.rs]: added

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 4]
```diff
@@ -60,6 +60,99 @@ impl ScalarFunc {
...
+    ) -> Result<Option<i64>> {
+        let substr = try_opt!(self.children[0].eval_string(ctx, row));
+        let s = try_opt!(self.children[1].eval_string(ctx, row));
+        let mut pos = try_opt!(self.children[2].eval_int(ctx, row));
+
+        pos -= 1; // transfer 1-based argument to 0-based real index
+        if pos < 0 {
+            return Ok(Some(0));
+        }
+        if pos > (s.len() as i64 - substr.len() as i64) {
```
[the remark this answers is not in the document]
Comment (gaodayue) [reply src/coprocessor/dag/expr/builtin_string.rs]: sure

[the remark this answers is not in the document]
Comment (gaodayue) [reply src/coprocessor/dag/expr/builtin_string.rs]: Cool

**Comment by gaodayue** [comment]
Thanks @rleungx @AndreMouche for reviewing. I've pushed code to address the last round of review comments, please check later.

**Pull request merged**